### DISCUSSION TITLE
Zod schema validation with union fails when schema has missing fields

### DISCUSSION THREAD

**Discussion by alveshelio** [colinhacks/zod#2977] [Q&A]
I'm having some a hard time understanding why my schema with union is failing.

Here's the schema: 
```ts
export const reportFormSchema = () => {
  const { validator } = reportTemplateNameValidator();

  const fieldSchema = z.object({
    id: z.nativeEnum(ReportFieldName),
    content: z.string().min(1),
    isTransferDisabled: z.boolean().optional(),
    isIndicator: z.boolean()
  });

  const reportFormSchemaBase = z.object({
    direction: z.nativeEnum(ReportDirection),
    format: z.nativeEnum(ReportFormat),
    type: z.nativeEnum(ReportType),
    fields: z
      .array(fieldSchema)
      .refine((items) => items.some((item) => item.isIndicator)),
    daysBeforeDepartures: z.array(z.string())
  });

  const reportTemplateFormSchema = reportFormSchemaBase.extend({
    name: z.string().trim().refine(validator, {
      message: "nameAlreadyExists"
    }),
    mode: z.literal(FORM_MODE.REPORT_TEMPLATE),
    scopeCode: z.undefined()
  });

  const reportGenerateFormSchema = reportFormSchemaBase.extend({
    name: z.string().trim().min(1, { message: "fieldRequired" }),
    mode: z.literal(FORM_MODE.REPORT_GENERATE),
    scopeCode: z.string().min(1)
  });

  return z
    .union([reportTemplateFormSchema, reportGenerateFormSchema])
    .refine((formData) =>
      formData.type === ReportType.OD_CLASS_DBD ||
      formData.type === ReportType.LEG_CLASS_DBD
        ? formData.daysBeforeDepartures.length > 0
        : formData.daysBeforeDepartures.length === 0
    );
};

export type ReportFormSchema = z.infer<ReturnType<typeof reportFormSchema>>;
```

I want `scopeCode` to be a `string` when in `FORM_MODE.REPORT_GENERATE` and `undefined` when in `FORM_MODE.REPORT_TEMPLATE`, however, when I try to parse this data:
```ts
const result = reportFormSchema();
    const parseResult = result.safeParse({
      scopeCode: "SC0000000432",
      mode: FORM_MODE.REPORT_GENERATE,
      daysBeforeDepartures: [],
    });
```
My test fails, which is expected since there are missing fields, however, `scopeCode` should not be on error since it is not `undefined` and I'm in `mode:  FORM_MODE.REPORT_GENERATE`, here's the error on field `scopeCode:
```ts
{
  "code": "invalid_type",
  "expected": "undefined",
  "received": "string",
  "path": [
    "scopeCode"
  ],
  "message": "Expected undefined, received string"
}
```
What's even stranger is that I get this error too:
```ts
{
  received: 'reportGenerate',
  code: 'invalid_literal',
  expected: 'reportTemplate',
  path: ['mode'],
  message: 'Invalid literal value, expected "reportTemplate"',
},
```
I'm not sure why I'm getting this error and I suspect that's why I'm getting the error on `scopeCode`.

When validating the schema with all the required data, the test passes and `scopeCode` isn't on error anymore.
```ts
const result = reportFormSchema();
    const parseResult = result.safeParse({
    direction: ReportDirection.BY_COLUMN,
    format: ReportFormat.XLSX,
    type: ReportType.WEEK_AGGR,
    name: "test",
    fields: [
      {
        id: ReportFieldName.DEPARTURE_DATE,
        content: "Operator",
        isTransferDisabled: false,
        isIndicator: true,

      {
        id: ReportFieldName.DEPARTURE_WEEK,
        content: "Departure Date",
        isTransferDisabled: false,
        isIndicator: false,
      },
    ],
      scopeCode: "SC0000000432",
      mode: FORM_MODE.REPORT_GENERATE,
      daysBeforeDepartures: [],
    });
```
I dont know what I'm doing wrong, from my understanding of how `union` works in Zod, the schema should be working.

I've created a [codesandbox](https://codesandbox.io/p/sandbox/blissful-rui-52hx66?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cloziuxm800073b6hbmdtfarm%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cloziuxm800033b6ha3ld1hnk%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cloziuxm800053b6hvjo15gfs%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cloziuxm800063b6hum4d05m0%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B81.36440013442366%252C18.63559986557634%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cloziuxm800033b6ha3ld1hnk%2522%253A%257B%2522id%2522%253A%2522cloziuxm800033b6ha3ld1hnk%2522%252C%2522activeTabId%2522%253A%2522clozj74u101gl3b6hng9icvjf%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cloziuxm800023b6hdfll1qbt%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Ftsconfig.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fjest.config.ts%2522%252C%2522id%2522%253A%2522clozixax400ae3b6h4hoz6izq%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.ts%2522%252C%2522id%2522%253A%2522clozj2xzu00n13b6hqi40tqie%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.test.ts%2522%252C%2522id%2522%253A%2522clozj74u101gl3b6hng9icvjf%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522cloziuxm800063b6hum4d05m0%2522%253A%257B%2522id%2522%253A%2522cloziuxm800063b6hum4d05m0%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522cloziuxm800053b6hvjo15gfs%2522%253A%257B%2522id%2522%253A%2522cloziuxm800053b6hvjo15gfs%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cloziuxm800043b6hiz1mj1h2%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cloziuxqe0009efd97phg3lq4%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cloziuxm800043b6hiz1mj1h2%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) with the schema and some testes, the one failing is `'should return a VALID SCHEMA object when report in "reportGenerate" mode and only "scopeCode" and "daysBeforeDepartures"'`

Any help would be appreciated.

**Comment by alveshelio** [comment]
Does anyone has any example on working with Zod union with partial valid schemas?
I'm really stuck here, I'm out of ideas. 

Thanks

**Reply by JacobWeisenburger** [reply]
What do you mean by partial valid schema?

**Comment by alveshelio** [comment]
So what I need is that when I'm in mode 'reportGenerate', the `scopeCode` get's populated. However, if the user changes the `scopeCode`,  then the schema should be invalid for all the other fields except for `scopeCode`, but that's not the case.  I get these errors:

```ts
{
  received: 'reportGenerate',
  code: 'invalid_literal',
  expected: 'reportTemplate',
  path: ['mode'],
  message: 'Invalid literal value, expected "reportTemplate"',
},
```
I really don't understand why I'm getting this error since I've set the `defaultValues` to be 
```ts
{
  mode: 'generateReport',
}
```

And this is probably why I'm then getting this error:
{
  "code": "invalid_type",
  "expected": "undefined",
  "received": "string",
  "path": [
    "scopeCode"
  ],
  "message": "Expected undefined, received string"
}

These errors only happens if only the `scopeCode` is filled and if I try to change the `scopeCode`, if the form has all fields filled up, then I don't get any error when changing `scopeCode`.

**Reply by JacobWeisenburger** [reply]
Please send a full reproducible code example.

![image](https://github.com/colinhacks/zod/assets/31667350/d2440733-04be-44d8-8723-8ce28f64f51c)
![image](https://github.com/colinhacks/zod/assets/31667350/d0240b10-e1e3-4329-874d-933fdb8d87f1)

**Reply by alveshelio** [reply]
Here's a [codesandbox](https://codesandbox.io/p/sandbox/blissful-rui-52hx66?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cloziuxm800073b6hbmdtfarm%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cloziuxm800033b6ha3ld1hnk%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cloziuxm800053b6hvjo15gfs%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cloziuxm800063b6hum4d05m0%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B81.36440013442366%252C18.63559986557634%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cloziuxm800033b6ha3ld1hnk%2522%253A%257B%2522id%2522%253A%2522cloziuxm800033b6ha3ld1hnk%2522%252C%2522activeTabId%2522%253A%2522clozj74u101gl3b6hng9icvjf%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cloziuxm800023b6hdfll1qbt%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Ftsconfig.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fjest.config.ts%2522%252C%2522id%2522%253A%2522clozixax400ae3b6h4hoz6izq%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.ts%2522%252C%2522id%2522%253A%2522clozj2xzu00n13b6hqi40tqie%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.test.ts%2522%252C%2522id%2522%253A%2522clozj74u101gl3b6hng9icvjf%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522cloziuxm800063b6hum4d05m0%2522%253A%257B%2522id%2522%253A%2522cloziuxm800063b6hum4d05m0%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522cloziuxm800053b6hvjo15gfs%2522%253A%257B%2522id%2522%253A%2522cloziuxm800053b6hvjo15gfs%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cloziuxm800043b6hiz1mj1h2%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cloziuxqe0009efd97phg3lq4%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cloziuxm800043b6hiz1mj1h2%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) with the schema and some testes, the one failing is 'should return a VALID SCHEMA object when report in "reportGenerate" mode and only "scopeCode" and "daysBeforeDepartures"'

**Reply by JacobWeisenburger** [reply]
unfortunately, this has a lot of problems that would take too long to diagnose for free. I'm happy to help, but it would require a [Sponsored Consulting Session](https://github.com/sponsors/JacobWeisenburger?frequency=one-time). Let me know if that's something you would be interested in.
